src/Microsoft.Health.Fhir.Api/Features/Context/FhirAuthenticationExceptionHandlerMiddleware.cs

```diff
@@ -0,0 +1,51 @@
+﻿// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System.Threading.Tasks;
+using EnsureThat;
+using Microsoft.AspNetCore.Http;
+using Microsoft.IdentityModel.S2S;
+using Microsoft.IdentityModel.Tokens;
+
+namespace Microsoft.Health.Fhir.Api.Features.Context
+{
+    /// <summary>
+    /// Middlware that runs before authentication middleware and after FhirRequestContextBeforeAuthenticationMiddleware
+    /// </summary>
+    public class FhirAuthenticationExceptionHandlerMiddleware
+    {
+        private readonly RequestDelegate _next;
+
+        public FhirAuthenticationExceptionHandlerMiddleware(
+            RequestDelegate next)
+        {
+            EnsureArg.IsNotNull(next, nameof(next));
+            _next = next;
+        }
+
+        public async Task Invoke(HttpContext context)
+        {
+            try
+            {
+                await _next(context);
+            }
+            catch (S2SAuthenticationException ex)
+            {
+                if (ex.InnerException is SecurityTokenInvalidAudienceException)
+                {
+                    throw new SecurityTokenInvalidAudienceException(ex.InnerException.Message, ex.InnerException);
+                }
+                else if (ex.InnerException is SecurityTokenInvalidIssuerException)
+                {
+                    throw new SecurityTokenInvalidIssuerException(ex.InnerException.Message, ex.InnerException);
+                }
+                else
+                {
+                    throw ex;
```

Comment:
I would replace "throw ex;" with "throw;".

"throw;" will preserve the original call stack.

Reply:
VS suggested that actually and I thought I changed it to just throw to keep the call stack. Good catch. Fixed it now.